modules/clightning-db-replication.nix

```diff
@@ -0,0 +1,223 @@
+    assertions = [
+      { assertion = useSshfs || (localDir != null);
+        message = ''
+          services.clightning.replication: One of `sshfs.destination` or
+          `local.directory` must be set.
+        '';
+      }
+      { assertion = !useSshfs || (localDir == null);
+        message = ''
+          services.clightning.replication: Only one of `sshfs.destination` and
+          `local.directory` must be set.
+        '';
+      }
+    ];
+
+    systemd.tmpfiles.rules = optional (localDir != null && cfg.local.setupDirectory)
+      "d '${localDir}' 0770 ${user} ${group} - -";
+
+    services.clightning.extraConfig = let
+      mainDB = "${clightning.dataDir}/${network}/lightningd.sqlite3";
+      replicaDB = "${destDir}/lightningd.sqlite3";
+    in ''
+      wallet=sqlite3://${mainDB}:${replicaDB}
+    '';
+
+    systemd.services.clightning = {
+      bindsTo = mkIf useMounts [ "clightning-replication-mounts.server" ];
+      serviceConfig.ReadWritePaths = [
+        # We can't simply set `destDir` here because it might point to
+        # a FUSE mount.
+        # FUSE mounts can only be set up as `ReadWritePaths` by systemd when they
+        # are accessible by root. This would require FUSE-mounting with option
+        # `allow_other`.
+        (if useMounts then mountsDir else localDir)
+      ];
+    };
+
+    systemd.services.clightning-replication-mounts = mkIf useMounts {
+      requiredBy = [ "clightning.service" ];
+      before = [ "clightning.service" ];
+      after = [ "setup-secrets.service" ];
+      path = [
+        # Includes
+        # - The SUID-wrapped `fusermount` binary which enables FUSE
+        #   for non-root users
+        # - The SUID-wrapped `mount` binary, used for unmounting
+        "/run/wrappers"
+      ] ++ optionals cfg.encrypt [
+        # Includes `logger`, required by gocryptfs
+        pkgs.util-linux
+      ];
+
+      script =
+        optionalString useSshfs ''
+          mkdir -p ${sshfsDir}
+          ${pkgs.sshfs}/bin/sshfs ${cfg.sshfs.destination} -p ${toString cfg.sshfs.port} ${sshfsDir} \
+            -o ${builtins.concatStringsSep "," ([
+              "IdentityFile='${secretsDir}'/clightning-replication-ssh-key"
+            ] ++ cfg.sshfs.sshOptions)}
+        '' +
+        optionalString cfg.encrypt ''
+          cipherDir="${if useSshfs then sshfsDir else localDir}/lightningd-db"
+          mkdir -p "$cipherDir" ${plaintextDir}
+          gocryptfs=(${pkgs.gocryptfs}/bin/gocryptfs -passfile '${secretsDir}'/clightning-replication-password)
+          # 1. init
+          if [[ ! -e $cipherDir/gocryptfs.conf ]]; then
+            "''${gocryptfs[@]}" -init "$cipherDir"
+          fi
+          # 2. mount
+          "''${gocryptfs[@]}" "$cipherDir" ${plaintextDir}
+        '';
+
+      preStop =
+        optionalString cfg.encrypt ''
+          umount ${plaintextDir} || true
+        '' +
+        optionalString useSshfs ''
+          umount ${sshfsDir}
+        '';
```

Comment:
What if the service aborts in a way that the preStop isn't executed? Should we try unmounting at service start?

Reply:
You mean (re-)unmounting so that the service will mount the directory fresh every time?

Reply:
> aborts in a way that the preStop isn't executed

We should use `postStop` instead. Barring a systemd bug or a crash of the systemd process (both very improbable), `postStop` is guaranteed to be executed after the service was started.
`preStop` wouldn't be called if `ExecStart` mounts a SSHFs but fails setting up the cryptfs, thus leaking a SSHFs mount after service stop.

Reply:
Implemented in ab9a38b4bd99aca00d1274eda2ea05f1dee41f2e